### ISSUE TITLE
test failures on SBCL and CMUCL

### ISSUE THREAD

**Issue by sellout** [sellout/quid-pro-quo#1]
`(asdf:test-system :quid-pro-quo)` fails a number of tests on SBCL 1.0.47 with “invalid number of arguments: 2”, but I can't decipher it (CMUCL gives a similar error). There must be something wrong with this part of metaclass.lisp:

``` lisp
(defun passes-invariants-p (object)
  (and (passes-slot-type-invariants-p object)
       (passes-class-invariants-p object)))

(defun add-reader-invariant (reader class)
  (add-method (ensure-generic-function reader
                                       :lambda-list '(object)
                                       :method-combination '(contract))
              (make-instance 'standard-method
                             :qualifiers '(invariant)
                             :lambda-list '(object)
                             :specializers (list class)
                             :function #'passes-invariants-p)))

(defun add-writer-invariant (writer class)
  (add-method (ensure-generic-function writer
                                       :lambda-list '(new-value object)
                                       :method-combination '(contract))
              (make-instance 'standard-method
                             :qualifiers '(invariant)
                             :lambda-list '(new-value object)
                             :specializers (list (find-class t) class)
                             :function (lambda (new-value object)
                                         (declare (ignore new-value))
                                         (passes-invariants-p object)))))
```

But it seems to me that `PASSES-INVARIANTS-P` always gets called with one argument.

And here's the stack:

```
  0: (QUID-PRO-QUO::PASSES-INVARIANTS-P (#<QUID-PRO-QUO-TEST::TEST-1 {1202EF91}>))[:EXTERNAL]
  1: ((SB-PCL::EMF QUID-PRO-QUO-TEST::MY-SLOT) #<unavailable argument> #<unavailable argument> #<QUID-PRO-QUO-TEST::TEST-1 {1202EF91}>)
  2: ((SB-PCL::FAST-METHOD QUID-PRO-QUO-TEST:TEST-QPQ :PRECONDITION "first arg needs non-zero my-slot" (QUID-PRO-QUO-TEST::TEST-1 QUID-PRO-QUO-TEST::TEST-1)) ..)
  3: ((SB-PCL::EMF QUID-PRO-QUO-TEST:TEST-QPQ) #<unavailable argument> #<unavailable argument> #<QUID-PRO-QUO-TEST::TEST-1 {1202EF91}> #<QUID-PRO-QUO-TEST::TEST-1 {1203F8E1}>)
  4: ((LAMBDA ()))
  5: ((LABELS IT.BESE.FIVEAM::RUN-IT))
  6: ((SB-PCL::FAST-METHOD IT.BESE.FIVEAM::RUN-TEST-LAMBDA (IT.BESE.FIVEAM::TEST-CASE)) ..)
  7: ((SB-PCL::FAST-METHOD IT.BESE.FIVEAM::RUN-RESOLVING-DEPENDENCIES (IT.BESE.FIVEAM::TEST-CASE)) ..)
  8: ((FLET IT.BESE.FIVEAM::RUN-TESTS))
  9: ((SB-PCL::FAST-METHOD IT.BESE.FIVEAM::%RUN (IT.BESE.FIVEAM::TEST-SUITE)) #<unavailable argument> #<unavailable argument> #<IT.BESE.FIVEAM::TEST-SUITE QUID-PRO-QUO-TEST::TESTS {12A77DD1}>)
 10: ((LAMBDA ()))
 11: (IT.BESE.FIVEAM:RUN! QUID-PRO-QUO-TEST::TESTS)
 12: ((SB-PCL::EMF ASDF:PERFORM) #<unavailable argument> #<unavailable argument> #<ASDF:TEST-OP NIL {133899B1}> #<ASDF:SYSTEM "quid-pro-quo">)
 13: ((LAMBDA ()))
 14: ((FLET SB-THREAD::WITH-RECURSIVE-LOCK-THUNK))
 15: ((FLET SB-C::WITH-IT))
 16: ((SB-PCL::FAST-METHOD ASDF:OPERATE (T T)) #<unused argument> #<unused argument> ASDF:TEST-OP "quid-pro-quo")
 17: ((SB-PCL::EMF ASDF:OPERATE) #<unused argument> #<unused argument> ASDF:TEST-OP "quid-pro-quo" 4717675 0)
 18: ((LAMBDA ()))
 19: ((FLET SWANK-BACKEND:CALL-WITH-COMPILATION-HOOKS) #<CLOSURE (LAMBDA #) {133896DD}>)
 --more--
```

**Comment by sellout** [comment]
Possibly related to https://bugs.launchpad.net/sbcl/+bug/767615 – waiting to find out.

**Comment by sellout** [comment]
```
; disassembly for QUID-PRO-QUO::PASSES-INVARIANTS-P
; 12AFC554:       8D5C24F8         LEA EBX, [ESP-8]           ; no-arg-parsing entry point
;       58:       83EC0C           SUB ESP, 12
;       5B:       8BD6             MOV EDX, ESI
;       5D:       8B0518C5AF12     MOV EAX, [#x12AFC518]      ; #<FDEFINITION object for QUID-PRO-QUO::PASSES-SLOT-TYPE-INVARIANTS-P>
;       63:       B904000000       MOV ECX, 4
;       68:       892B             MOV [EBX], EBP
;       6A:       8BEB             MOV EBP, EBX
;       6C:       FF5005           CALL DWORD PTR [EAX+5]
;       6F:       7302             JNB L0
;       71:       8BE3             MOV ESP, EBX
;       73: L0:   8B75FC           MOV ESI, [EBP-4]
;       76:       81FA0B001004     CMP EDX, 68157451
;       7C:       7525             JNE L3
;       7E: L1:   81FA0B001004     CMP EDX, 68157451
;       84:       750A             JNE L2
;       86:       BA0B001004       MOV EDX, 68157451
;       8B:       8BE5             MOV ESP, EBP
;       8D:       F8               CLC
;       8E:       5D               POP EBP
;       8F:       C3               RET
;       90: L2:   8BD6             MOV EDX, ESI
;       92:       8B051CC5AF12     MOV EAX, [#x12AFC51C]      ; #<FDEFINITION object for QUID-PRO-QUO::PASSES-CLASS-INVARIANTS-P>
;       98:       B904000000       MOV ECX, 4
;       9D:       FF7504           PUSH DWORD PTR [EBP+4]
;       A0:       FF6005           JMP DWORD PTR [EAX+5]
;       A3: L3:   81FA27001004     CMP EDX, 68157479
;       A9:       74D3             JEQ L1
;       AB:       8B0520C5AF12     MOV EAX, [#x12AFC520]      ; '(MEMBER
                                                              ;   NIL T)
;       B1:       0F0B0A           BREAK 10                   ; error trap
;       B4:       03               BYTE #X03
;       B5:       1F               BYTE #X1F                  ; OBJECT-NOT-TYPE-ERROR
;       B6:       90               BYTE #X90                  ; EDX
;       B7:       10               BYTE #X10                  ; EAX
;       B8:       0F0B0A           BREAK 10                   ; error trap
;       BB:       02               BYTE #X02
;       BC:       18               BYTE #X18                  ; INVALID-ARG-COUNT-ERROR
;       BD:       4F               BYTE #X4F                  ; ECX
```

**Comment by sellout** [comment]
Fixed in 3b8724eaeab7945fe3cf9d3a76123d085d2c090b, thanks to Krystof explaining how the MOP is supposed to work.
